tests/system/Database/Live/UpsertTest.php

```diff
@@ -0,0 +1,391 @@
+use stdclass;
+use Tests\Support\Database\Seeds\CITestSeeder;
+
+/**
+ * @group DatabaseLive
+ *
+ * @internal
+ */
+final class UpsertTest extends CIUnitTestCase
+{
+    use DatabaseTestTrait;
+
+    protected $refresh = true;
+    protected $seed    = CITestSeeder::class;
+
+    public function testUpsertOnUniqueIndex()
+    {
+        $userData = [
+            'email'   => 'upsertone@test.com',
+            'name'    => 'Upsert One',
+            'country' => 'US',
+        ];
+
+        $this->db->table('user')->upsert($userData);
+
+        $this->seeInDatabase('user', ['name' => 'Upsert One']);
+    }
+
+    public function testUpsertAndUpsertBatchWithObject()
+    {
+        $data = [];
+
+        $row          = new stdclass();
+        $row->name    = 'Pedro';
+        $row->email   = 'pedro@acme.com';
+        $row->country = 'El Salvador';
+        $data[]       = $row;
+
+        $row          = new stdclass();
+        $row->name    = 'Derek Jones';
+        $row->email   = 'derek@world.com';
+        $row->country = 'Canada';
+        $data[]       = $row;
+
+        $this->db->table('user')->upsertBatch($data);
+
+        $this->seeInDatabase('user', ['name' => 'Pedro']);
+        $this->seeInDatabase('user', ['name' => 'Derek Jones', 'country' => 'Canada']);
+
+        $row->country = 'Spain';
+        $this->db->table('user')->upsert($row);
+
+        $this->seeInDatabase('user', ['name' => 'Derek Jones', 'country' => 'Spain']);
+    }
+
+    public function testUpsertChangePrimaryKeyOnUniqueIndex()
+    {
+        $userData = [
+            'id'      => 5,
+            'email'   => 'ahmadinejad@world.com',
+            'name'    => 'Ahmadinejad',
+            'country' => 'Iran',
+        ];
+
+        // some databases don't allow update of auto incrememnt
+        if ($this->db->DBDriver === 'SQLSRV') {
+            $this->assertTrue(true);
+        } else {
+            $this->db->table('user')
+                ->onConstraint('email')
+                ->updateFields('id, name, country')
+                ->upsert($userData);
+
+            $new = $this->db->table('user')
+                ->getwhere(['email' => 'ahmadinejad@world.com'])
+                ->getRow();
+
+            $this->assertSame(5, (int) $new->id);
+        }
+    }
+
+    public function testGetCompiledUpsert()
+    {
+        switch ($this->db->DBDriver) {
+
+            case 'MySQLi':
+                $expected = <<<'SQL'
+                    INSERT INTO `db_user` (`country`, `email`, `name`)
+                    VALUES ('Iran','ahmadinejad@world.com','Ahmadinejad')
+                    ON DUPLICATE KEY UPDATE
+                    `country` = VALUES(`country`),
+                    `email` = VALUES(`email`),
+                    `name` = VALUES(`name`)
+                    SQL;
+                break;
+
+            case 'Postgre':
+                $expected = <<<'SQL'
+                    INSERT INTO "db_user"("country", "email", "name")
```

Comment:
There should be a space.

```suggestion
                    INSERT INTO "db_user" ("country", "email", "name")
```